### ISSUE TITLE
Support Python 3.12

### ISSUE THREAD

**Issue by EwoutH** [vllm-project/vllm#1218]
[Python 3.12](https://docs.python.org/3.12/whatsnew/3.12.html) releases next week, [Monday 2023-10-02](https://peps.python.org/pep-0693/). It would be great if vLLM could support Python 3.12 fully, including testing in CI and wheels uploaded [to PyPI](https://pypi.org/project/vllm/#files).

<details><summary>Summary – Python 3.12 Release highlights</summary>

> New grammar features:
> - [PEP 701: Syntactic formalization of f-strings](https://docs.python.org/3.12/whatsnew/3.12.html#whatsnew312-pep701)
> 
> Interpreter improvements:
> - [PEP 684: A Per-Interpreter GIL](https://docs.python.org/3.12/whatsnew/3.12.html#whatsnew312-pep684)
> - [PEP 669: Low impact monitoring for CPython](https://docs.python.org/3.12/whatsnew/3.12.html#whatsnew312-pep669)
> 
> New typing features:
> - [PEP 688: Making the buffer protocol accessible in Python](https://docs.python.org/3.12/whatsnew/3.12.html#whatsnew312-pep688)
> -  [PEP 692: Using TypedDict for more precise **kwargs typing](https://docs.python.org/3.12/whatsnew/3.12.html#whatsnew312-pep692)
> -  [PEP 695: Type Parameter Syntax](https://docs.python.org/3.12/whatsnew/3.12.html#whatsnew312-pep695)
> - [PEP 698: Override Decorator for Static Typing](https://docs.python.org/3.12/whatsnew/3.12.html#whatsnew312-pep698)
> 
> Important deprecations, removals or restrictions:
> -  [PEP 623](https://peps.python.org/pep-0623/): Remove wstr from Unicode
> -  [PEP 632](https://peps.python.org/pep-0632/): Remove the distutils package. See [the migration guide](https://peps.python.org/pep-0632/#migration-advice) for advice on its replacement.
> -  [gh-95299](https://github.com/python/cpython/issues/95299): Do not pre-install setuptools in virtual environments created with [venv](https://docs.python.org/3.12/library/venv.html#module-venv). This means that distutils, setuptools, pkg_resources, and easy_install will no longer available by default; to access these run pip install setuptools in the [activated](https://docs.python.org/3.12/library/venv.html#venv-explanation) virtual environment.</details>

**Comment by WoosukKwon** [comment]
Hi @EwoutH, thanks for bringing this up. Once our dependencies (e.g., PyTorch, Ray) support python 3.12, we will support it as well.

**Comment by kannon92** [comment]
Ray still seems to be not supporting python 3.12 yet: https://github.com/ray-project/ray/issues/40211

**Comment by greg2451** [comment]
Two days ago, PR to support 3.12 for ray was merged: https://github.com/ray-project/ray/pull/45621. We can hope support will arrive soon enough.

But as mentioned: 

> TL;DR; this PR will support python 3.12 as a nightly wheel; other release artifacts will take more work

**Comment by beniz** [comment]
Modifying `CMakeLists.txt` for Python 3.12 works fine:

```
diff --git a/CMakeLists.txt b/CMakeLists.txt
index aa15b632..a930f13e 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -13,7 +13,7 @@ include(${CMAKE_CURRENT_LIST_DIR}/cmake/utils.cmake)
 # Supported python versions.  These versions will be searched in order, the
 # first match will be selected.  These should be kept in sync with setup.py.
 #
-set(PYTHON_SUPPORTED_VERSIONS "3.8" "3.9" "3.10" "3.11")
+set(PYTHON_SUPPORTED_VERSIONS "3.8" "3.9" "3.10" "3.11" "3.12")

 # Supported NVIDIA architectures.
 set(CUDA_SUPPORTED_ARCHS "7.0;7.5;8.0;8.6;8.9;9.0")
```

**Comment by rockwotj** [comment]
I was able to build for python 3.12 as well

**Comment by ruiliuacd** [comment]
look forward to supporting python 3.13
